File: src/main/java/org/cimsbioko/search/SearchJob.java
package org.cimsbioko.search;

import android.util.Log;

import org.apache.lucene.search.IndexSearcher;

public abstract class SearchJob implements Runnable {

    private static final String TAG = SearchJob.class.getSimpleName();

    private SearchQueue service;

    void setQueue(SearchQueue service) {
        this.service = service;
    }

    private boolean isActive() {
        return service != null && !service.isShutdown();
    }

    @Override
    public void run() {
        if (isActive()) {
            SearchIndex index = SearchIndex.getInstance();
            try {
                IndexSearcher searcher = index.acquire();
                try {
                    performSearch(searcher);
                    if (isActive()) {
                        postResult();
                    }
                } finally {
                    index.release(searcher);
                }
            } catch (Exception e) {
                if (isActive()) {
                    handleException(e);
                }
            }
        }
    }

    public abstract void performSearch(IndexSearcher searcher) throws Exception;

    protected abstract void postResult() throws Exception;

    public void handleException(Exception e) {
        Log.e(TAG, "search job failed", e);
    }
}

File: src/main/kotlin/org/cimsbioko/search/SearchJob.kt
package org.cimsbioko.search

import android.util.Log
import org.apache.lucene.search.IndexSearcher
import org.cimsbioko.search.SearchIndex.acquire
import org.cimsbioko.search.SearchIndex.release

abstract class SearchJob : Runnable {

    internal var service: SearchQueue? = null

    private val isActive: Boolean
        get() = !(service?.isShutdown ?: true)

    override fun run() {
        if (isActive) {
            try {
                with(acquire()) {
                    try {
                        performSearch()
                        if (isActive) {
                            postResult()
                        }
                    } finally {
                        release()
                    }
                }
            } catch (e: Exception) {
                if (isActive) {
                    handleException(e)
                }
            }
        }
    }

    @Throws(Exception::class)
    abstract fun IndexSearcher.performSearch()

    @Throws(Exception::class)
    protected abstract fun postResult()

    open fun handleException(e: Exception) {
        Log.e(TAG, "search job failed", e)
    }

    companion object {
        private val TAG = SearchJob::class.java.simpleName
    }
}